witnet.toml

```diff
@@ -96,6 +96,8 @@ genesis_path = ".witnet/config/genesis_block.json"
 #mint_external_percentage = 50
 # Set a minimum fee you require before your node includes a value transfer transaction into a block
 minimum_vtt_fee_nanowits = 1
+# Set a minimum data request reward collateral ratio you expect before your node includes a data request into a block
+minimum_reward_collateral_ratio = 125
```

Comment:
If this is exactly what this does, I can't see the point.

Imagine that you decrease this value and hence your node becomes more picky, refusing to mine under-rewarding data requests. Rest assured that someone else will leave it to a default value, letting the data request in. Your node will still participate in resolving the under-rewarding data request that you censored before, right?

Reply:
I think it would make more sense to have a setting that makes your node refrain from committing to under-rewarding data requests as a mean for witnessing risk management.

Then the fact that anyone can set that value to whatever they want would obviously mess with the replication factor and potentially end up in `InsufficientCommits`,  but idk maybe we're ok with that because the automated mechanism for adjusting the witnessing difficulty seems to work like a charm rn.

Reply:
One way or another, let me advocate for protocol-level consensus constants over user-customizable settings. Every time we introduce a user-customizable setting, we're making things worse for the main two protocol personas:
- **Miners**: how can they assess what's a reasonable value to use for this setting? Are we explaining clearly enough that the bigger the number used here, the bigger the risk they take? Or that we can't go higher than 125 nor lower than 1? (of course that means that they can't require data requests to reward more than the collateral amount)
- **Requesters**: same story here. If we let miners reject or delay transactions based on this setting, then the request-to-collateral ratio becomes non-trivial for the requester. How can they figure out what's a reasonable ratio? Why would they not use the default value, provided that this won't affect transaction priority? What does this number mean to the requester at all, given that they're not the ones assessing risk? How will they understand that they need to decrease their ratio if a data request errors because of a legal ratio (<=125) but still not good enough for a majority of miners?

Reply:
Summing up, you know I'm all for streamlining the user experience of Witnet, and I think that having this particular user setting works radically against that.

Reply:
![image](https://user-images.githubusercontent.com/1316439/214554035-e1dacc94-7f47-4145-8587-49e42ec6cbe3.png)

Reply:
> Rest assured that someone else will leave it to a default value, letting the data request in. Your node will still participate in resolving the under-rewarding data request that you censored before, right?

Correct on both accounts. When creating the reference implementation, I did not realize that you will still solve the request after someone else has included it, which makes that setting next to useless.

> I think it would make more sense to have a setting that makes your node refrain from committing to under-rewarding data requests as a mean for witnessing risk management.

That's the general idea I had in mind, yes. Adding a setting like that offered an easy-to-use solution for miners to 'disagree' with the protocol-set risk-rewards parameter. Given that anyone could easily modify his nodes to do this, I saw little harm in adding it out-of-the-box.

> Then the fact that anyone can set that value to whatever they want would obviously mess with the replication factor and potentially end up in InsufficientCommits, but idk maybe we're ok with that because the automated mechanism for adjusting the witnessing difficulty seems to work like a charm rn.

Given that the replication factor will result in 16 times the number of required witnesses, I doubt we'll quickly see InsufficientCommits data requests due to this additional setting. If we do, that means a very large majority of miners has decided it does not want to solve data requests at a certain ratio. And then we arrive at below point:

> how can they assess what's a reasonable value to use for this setting

I'd argue both miners and requesters can relatively easily assess this by looking at the data requests flowing network. If a majority of data requests are getting solved with a specific ratio, all participants know a high amount of miners consider that setting to be acceptable. It then seems like a relatively straightfoward conclusion that electing to set a lower ratio means requesters risk the data request not being solved and miners risk not solving them thereby losing reputation.

You can probably make the point that this may result in requesters overpaying for solving the data requests (akin to how gas pricing used to work in Ethereum before EIP-1559). However, I'd think a prediction of what a reasonable value for parameter should be, is something that could easily be included in external tooling.

I do follow your point that, overall, it may decrease the UX to the point that it is discutable whether this setting offers a significant enough advantage to warrant its existence. I have no strong preference on keeping this in or simply removing it.

Reply:
So I don't see any problem with that config parameter because it is optional, so miners aren't forced to choose a value, and also because we cannot forbid miners from implementing custom rules to decide which transactions to include into blocks.

It is true that it is a very bad user experience to have a transaction delayed or rejected for an unknown reason, but that's the nature of decentralized blockchains.

I can try to add another config parameter related to witnessing risk management, to allow node operators to choose an acceptable reward-collateral ratio before committing to a data request. Keep in mind that these config parameters can be implemented without breaking changes, so we can always easily implement them in the future.

Reply:
> Keep in mind that these config parameters can be implemented without breaking changes, so we can always easily implement them in the future.

Is it OK if we ditch these setting for now and then introduce it if there's such a clear feature request from the node operators community? At that point we will also have more info on how the fixed ratio plays out, so as to make a better decision.

Reply:
I don't see any problem with keeping it, but I pushed a commit that removes it.

Reply:
Is that a temporary commit? Right now this branch contains unused functions / variables / tests whichs makes it somewhat hard to understand?

Reply:
No, it is not temporary. Which functions are unused? The compiler should have complained about them.

Reply:
`set_minimum_reward_collateral_ratio` is not used anywhere anymore except in tests? I guess I'm a bit confused why we still use the `minimum_reward_collateral_ratio` variable when we could just use `required_reward_collateral_ratio`? Seems cleaner to me to reformat it?

Reply:
You are right, removed `set_minimum_reward_collateral_ratio` and its tests in the latest commit.